In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [2]:
from sklearn.datasets import load_breast_cancer
d = load_breast_cancer()

In [3]:
data =d['data']

In [4]:
data = pd.DataFrame(data)

In [5]:
data.loc[0]

0       17.990000
1       10.380000
2      122.800000
3     1001.000000
4        0.118400
5        0.277600
6        0.300100
7        0.147100
8        0.241900
9        0.078710
10       1.095000
11       0.905300
12       8.589000
13     153.400000
14       0.006399
15       0.049040
16       0.053730
17       0.015870
18       0.030030
19       0.006193
20      25.380000
21      17.330000
22     184.600000
23    2019.000000
24       0.162200
25       0.665600
26       0.711900
27       0.265400
28       0.460100
29       0.118900
Name: 0, dtype: float64

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 30 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       569 non-null    float64
 1   1       569 non-null    float64
 2   2       569 non-null    float64
 3   3       569 non-null    float64
 4   4       569 non-null    float64
 5   5       569 non-null    float64
 6   6       569 non-null    float64
 7   7       569 non-null    float64
 8   8       569 non-null    float64
 9   9       569 non-null    float64
 10  10      569 non-null    float64
 11  11      569 non-null    float64
 12  12      569 non-null    float64
 13  13      569 non-null    float64
 14  14      569 non-null    float64
 15  15      569 non-null    float64
 16  16      569 non-null    float64
 17  17      569 non-null    float64
 18  18      569 non-null    float64
 19  19      569 non-null    float64
 20  20      569 non-null    float64
 21  21      569 non-null    float64
 22  22

In [7]:
data.describe().T

count        mean         std         min         25%         50%  \
0   569.0   14.127292    3.524049    6.981000   11.700000   13.370000   
1   569.0   19.289649    4.301036    9.710000   16.170000   18.840000   
2   569.0   91.969033   24.298981   43.790000   75.170000   86.240000   
3   569.0  654.889104  351.914129  143.500000  420.300000  551.100000   
4   569.0    0.096360    0.014064    0.052630    0.086370    0.095870   
5   569.0    0.104341    0.052813    0.019380    0.064920    0.092630   
6   569.0    0.088799    0.079720    0.000000    0.029560    0.061540   
7   569.0    0.048919    0.038803    0.000000    0.020310    0.033500   
8   569.0    0.181162    0.027414    0.106000    0.161900    0.179200   
9   569.0    0.062798    0.007060    0.049960    0.057700    0.061540   
10  569.0    0.405172    0.277313    0.111500    0.232400    0.324200   
11  569.0    1.216853    0.551648    0.360200    0.833900    1.108000   
12  569.0    2.866059    2.021855    0.757000    1.606000    2.287000   
13  569.0   40.337079   45.491006    6.802000   17.850000   24.530000   
14  569.0    0.007041    0.003003    0.001713    0.005169    0.006380   
15  569.0    0.025478    0.017908    0.002252    0.013080    0.020450   
16  569.0    0.031894    0.030186    0.000000    0.015090    0.025890   
17  569.0    0.011796    0.006170    0.000000    0.007638    0.010930   
18  569.0    0.020542    0.008266    0.007882    0.015160    0.018730   
19  569.0    0.003795    0.002646    0.000895    0.002248    0.003187   
20  569.0   16.269190    4.833242    7.930000   13.010000   14.970000   
21  569.0   25.677223    6.146258   12.020000   21.080000   25.410000   
22  569.0  107.261213   33.602542   50.410000   84.110000   97.660000   
23  569.0  880.583128  569.356993  185.200000  515.300000  686.500000   
24  569.0    0.132369    0.022832    0.071170    0.116600    0.131300   
25  569.0    0.254265    0.157336    0.027290    0.147200    0.211900   
26  569.0    0.272188    0.208624    0.000000    0.114500    0.226700   
27  569.0    0.114606    0.065732    0.000000    0.064930    0.099930   
28  569.0    0.290076    0.061867    0.156500    0.250400    0.282200   
29  569.0    0.083946    0.018061    0.055040    0.071460    0.080040   

            75%         max  
0     15.780000    28.11000  
1     21.800000    39.28000  
2    104.100000   188.50000  
3    782.700000  2501.00000  
4      0.105300     0.16340  
5      0.130400     0.34540  
6      0.130700     0.42680  
7      0.074000     0.20120  
8      0.195700     0.30400  
9      0.066120     0.09744  
10     0.478900     2.87300  
11     1.474000     4.88500  
12     3.357000    21.98000  
13    45.190000   542.20000  
14     0.008146     0.03113  
15     0.032450     0.13540  
16     0.042050     0.39600  
17     0.014710     0.05279  
18     0.023480     0.07895  
19     0.004558     0.02984  
20    18.790000    36.04000  
21    29.720000    49.54000  
22   125.400000   251.20000  
23  1084.000000  4254.00000  
24     0.146000     0.22260  
25     0.339100     1.05800  
26     0.382900     1.25200  
27     0.161400     0.29100  
28     0.317900     0.66380  
29     0.092080     0.20750

In [8]:
data.isna().any()

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
dtype: bool

In [9]:
d.feature_names

array(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error',
       'fractal dimension error', 'worst radius', 'worst texture',
       'worst perimeter', 'worst area', 'worst smoothness',
       'worst compactness', 'worst concavity', 'worst concave points',
       'worst symmetry', 'worst fractal dimension'], dtype='<U23')

In [10]:
d.target_names

array(['malignant', 'benign'], dtype='<U9')

In [11]:
output = d.target.reshape(-1,1)
output.shape

(569, 1)

In [12]:
output.shape

(569, 1)

In [13]:
from sklearn.model_selection import train_test_split
xtrain, xtest, ytrain, ytest =  train_test_split(data,output, test_size=0.25, random_state=31)

In [14]:
xtrain.shape

(426, 30)

In [15]:
xtest.shape

(143, 30)

In [16]:
ytrain.shape

(426, 1)

In [17]:
ytest.shape

(143, 1)

In [18]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
xtrain = sc.fit_transform(xtrain)

In [19]:
xtrain

array([[-0.81425697, -1.4355008 , -0.77527732, ...,  0.27034015,
         0.67335426,  0.44999932],
       [-0.34270873, -0.6518781 , -0.38453948, ..., -0.47543235,
         0.17465243, -0.77230978],
       [-0.86888756, -0.07055563, -0.76943919, ...,  2.08915035,
         1.94121376,  3.04727131],
       ...,
       [ 1.61824185,  0.15964807,  1.62419498, ...,  2.06351206,
        -0.25526294,  2.47549389],
       [ 0.34448656,  1.42693106,  0.44405827, ...,  1.3350831 ,
         1.74267102,  0.80655109],
       [ 0.16621832,  0.23173206,  0.12462905, ...,  0.70920729,
         0.18246907, -0.08887393]])

In [20]:
xtest = sc.fit_transform(xtest)

In [21]:
xtest

array([[ 4.68870110e-03,  1.90309232e+00, -2.68518559e-03, ...,
        -5.43079626e-01, -1.78317069e+00, -3.88159653e-01],
       [-2.72393582e-01,  1.29184322e+00, -3.38917726e-01, ...,
        -8.90628545e-01, -7.71437023e-01, -9.99744995e-01],
       [-8.30997449e-02, -1.21082559e-01, -1.54686168e-01, ...,
        -5.53911097e-01, -9.32145619e-01, -5.05258830e-01],
       ...,
       [-7.96380869e-01, -3.67933154e-01, -7.43590499e-01, ...,
         7.87095627e-02,  8.06429200e-01,  2.75106908e+00],
       [ 4.18940431e-01, -1.99244516e+00,  4.05170854e-01, ...,
         4.46037688e-01, -5.65072575e-01, -1.54568029e-01],
       [-3.71155584e-01,  3.79671504e-01, -3.97410250e-01, ...,
        -9.54047588e-01, -6.91082724e-01, -1.72769974e-01]])

In [22]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()
lr.fit(xtrain, ytrain)

LogisticRegression()

In [23]:
ypred = lr.predict(xtest)

In [24]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

In [25]:
print(accuracy_score(ytest, ypred))

0.9790209790209791


In [26]:
print(confusion_matrix(ytest, ypred))

[[53  3]
 [ 0 87]]


In [27]:
print(classification_report(ytest, ypred))

              precision    recall  f1-score   support

           0       1.00      0.95      0.97        56
           1       0.97      1.00      0.98        87

    accuracy                           0.98       143
   macro avg       0.98      0.97      0.98       143
weighted avg       0.98      0.98      0.98       143



In [28]:
from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()
gnb.fit(xtrain, ytrain)

GaussianNB()

In [29]:
ypred_gnb = gnb.predict(xtest)

In [30]:
print(accuracy_score(ytest, ypred_gnb))

0.9090909090909091


In [31]:
print(confusion_matrix(ytest, ypred_gnb))

[[48  8]
 [ 5 82]]


In [32]:
print(classification_report(ytest, ypred_gnb))

              precision    recall  f1-score   support

           0       0.91      0.86      0.88        56
           1       0.91      0.94      0.93        87

    accuracy                           0.91       143
   macro avg       0.91      0.90      0.90       143
weighted avg       0.91      0.91      0.91       143



In [33]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(xtrain, ytrain)

KNeighborsClassifier()

In [34]:
ypred_knn = knn.predict(xtest)

In [35]:
print(accuracy_score(ytest, ypred_knn))

0.958041958041958


In [36]:
print(confusion_matrix(ytest, ypred_knn))

[[51  5]
 [ 1 86]]


In [37]:
from sklearn.metrics import accuracy_score
for i in range(2,21):
    knn = KNeighborsClassifier(n_neighbors=i)
    knn.fit(xtrain, ytrain)
    
    ypred_knn = knn.predict(xtest)
    acc = accuracy_score(ytest, ypred_knn)
    print('{:20} {}'.format(acc, i))

   0.958041958041958 2
   0.958041958041958 3
   0.951048951048951 4
   0.958041958041958 5
   0.951048951048951 6
  0.9440559440559441 7
   0.965034965034965 8
   0.958041958041958 9
   0.965034965034965 10
  0.9440559440559441 11
  0.9440559440559441 12
  0.9440559440559441 13
   0.951048951048951 14
  0.9440559440559441 15
   0.958041958041958 16
  0.9300699300699301 17
  0.9300699300699301 18
  0.9300699300699301 19
  0.9440559440559441 20


In [38]:
#knn 96.50 %  when k =8

In [39]:
from sklearn.svm import SVC
svc = SVC(kernel='rbf')
svc.fit(xtrain, ytrain)

SVC()

In [40]:
ypred_svc = svc.predict(xtest)

In [41]:
print(accuracy_score(ytest, ypred_svc))

0.965034965034965


In [42]:
print(confusion_matrix(ytest, ypred_svc))

[[52  4]
 [ 1 86]]


In [43]:
from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier(criterion='entropy', n_estimators= 25)
rfc.fit(xtrain, ytrain)

RandomForestClassifier(criterion='entropy', n_estimators=25)

In [44]:
ypred_rfc = rfc.predict(xtest)

In [45]:
print(accuracy_score(ytest, ypred_rfc))

0.958041958041958


In [46]:
for i in range(2,51):
    rfc = RandomForestClassifier(criterion='entropy', n_estimators= i)
    rfc.fit(xtrain, ytrain)
    
    ypred_rfc = rfc.predict(xtest)
    
    acc = accuracy_score(ytest, ypred_rfc)
    print('{:20} {}'.format(acc, i))

  0.9020979020979021 2
  0.9300699300699301 3
  0.9440559440559441 4
  0.9440559440559441 5
  0.9230769230769231 6
  0.9370629370629371 7
  0.9440559440559441 8
  0.9370629370629371 9
  0.9440559440559441 10
   0.958041958041958 11
   0.958041958041958 12
  0.9300699300699301 13
  0.9370629370629371 14
  0.9370629370629371 15
  0.9370629370629371 16
  0.9440559440559441 17
  0.9370629370629371 18
  0.9370629370629371 19
   0.958041958041958 20
  0.9440559440559441 21
   0.965034965034965 22
  0.9370629370629371 23
   0.951048951048951 24
  0.9370629370629371 25
   0.951048951048951 26
   0.951048951048951 27
  0.9230769230769231 28
  0.9440559440559441 29
  0.9370629370629371 30
  0.9370629370629371 31
   0.958041958041958 32
  0.9440559440559441 33
   0.951048951048951 34
   0.958041958041958 35
   0.965034965034965 36
  0.9370629370629371 37
   0.951048951048951 38
  0.9300699300699301 39
   0.951048951048951 40
   0.951048951048951 41
   0.965034965034965 42
   0.972027972027972 43


In [47]:
# n_estimators= 30 , acc= 97.20 %

## Conclusion

Logistic Regression giving Highest accuracy with 97.90% ,  
Random Forest with the Second Highest accuracy of 97.20%

# Testing

In [48]:
sample1 = data.iloc[:1,:].values   #it has output 0
sample2 = data.iloc[18:19, :].values  #it has output 1
sample3 = data.iloc[19:20, :].values  # it has output 1
sample4 = data.iloc[20:21, :].values  # it has output 1


In [49]:
print(sc.transform([[1.799e+01, 1.038e+01, 1.228e+02, 1.001e+03, 1.184e-01, 2.776e-01,
        3.001e-01, 1.471e-01, 2.419e-01, 7.871e-02, 1.095e+00, 9.053e-01,
        8.589e+00, 1.534e+02, 6.399e-03, 4.904e-02, 5.373e-02, 1.587e-02,
        3.003e-02, 6.193e-03, 2.538e+01, 1.733e+01, 1.846e+02, 2.019e+03,
        1.622e-01, 6.656e-01, 7.119e-01, 2.654e-01, 4.601e-01, 1.189e-01]]))

[[ 1.03894633 -2.20403139  1.20098752  0.92543369  1.73640664  3.44137777
   2.57245119  2.60201424  2.62955796  2.37113699  3.19255143 -0.63326262
   3.65771605  3.63706885 -0.22762484  1.3744976   0.49527814  0.53917074
   1.40889829  0.74692198  1.88532287 -1.48519381  2.30180182  2.05003903
   1.42594285  2.86026266  2.15071358  2.33291122  3.16774983  2.09579907]]


In [50]:
sample2

array([[1.981e+01, 2.215e+01, 1.300e+02, 1.260e+03, 9.831e-02, 1.027e-01,
        1.479e-01, 9.498e-02, 1.582e-01, 5.395e-02, 7.582e-01, 1.017e+00,
        5.865e+00, 1.124e+02, 6.494e-03, 1.893e-02, 3.391e-02, 1.521e-02,
        1.356e-02, 1.997e-03, 2.732e+01, 3.088e+01, 1.868e+02, 2.398e+03,
        1.512e-01, 3.150e-01, 5.372e-01, 2.388e-01, 2.768e-01, 7.615e-02]])

In [51]:
print(lr.predict(sc.transform(sample1)))

[0]


In [52]:
print(lr.predict(sc.transform(sample2)))

[0]


In [53]:
print(lr.predict(sc.transform(sample3)))

[1]


In [54]:
print(lr.predict(sc.transform(sample4)))

[1]


###    Sample no 2 wrong classified by logistic reg model while sample 1, 3, 4 correctly classified